### PR TITLE
Document changes to writing DDR extensions

### LINKED ISSUE #1306: Document changes to writing DDR extensions
**Issue or pull request number:**
https://github.com/eclipse-openj9/openj9/pull/19173

**Overview:**
The names used in DDR code to refer to fields of `J9BuildFlags` have changed - they now consistently use the names of macros appearing in `j9cfg.h` (e.g. `env_data64` must now be referred to as `J9VM_ENV_DATA64`).

**Release target:**
*Eclipse OpenJ9 0.44.0*

**Applies to the following JDK versions:**
All JDK versions.

**Applies to the following platforms:**
All platforms.

### PR DESCRIPTION
**Pull request by Sreekala-Gopakumar**
https://github.com/eclipse-openj9/openj9-docs/issues/1306

The change is explained in the What's new in version 0.44.0 topic.

Closes #1306
Signed-off-by: Sreekala Gopakumar sreekala.gopakumar@ibm.com

### CODE REVIEW

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@keithc-ca @pshipton - Could you please check. Thanks!

There are some doubts as well as mentioned in https://github.com/eclipse-openj9/openj9-docs/issues/1306#issuecomment-2025777462.

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@pshipton @keithc-ca - I have updated the content. Please check and let me know if any further changes are required. 

Also, @keithc-ca - Is there any other simpler way to say "...raw name of the flag names"? What do you mean by "raw name"?

Thanks!

File: docs/version0.44.md [hunk 1]
```diff
@@ -76,6 +77,12 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
 
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Change in the Direct Dump Reader plug-ins code
```
Comment (pshipton) [inline docs/version0.44.md:80]: I suggest "the flag names that are used in the DDR code to refer to the fields of `J9BuildFlags` are changed" or similar be the title of the What's New. User's shouldn't have to read the big description to figure out what changed.

File: docs/version0.44.md [hunk 2]
```diff
@@ -76,6 +77,12 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
...
+
+Now, the flag names that are used in the DDR code to refer to the fields of `J9BuildFlags` are changed. The flag names are now those names that appear in C or C++ code, independent of the VM that produced a system dump. To reduce complexity and improve the uniformity of how the DDR code uses the flags, you must use the names that are specified in `j9cfg.h` as flag names in the plug-in code. For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`.
```
Comment (pshipton) [inline docs/version0.44.md:84]: I suggest this paragraph go first, followed by the lengthy description. The description is only needed for those who don't already understand, or those who want more details.

Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:84]: Usually, the strategy that we use is to explain the problem first and then explain the solution or the change that was made to mitigate that problem. Maybe I can try and reduce the first paragraph?

Comment (pshipton) [reply docs/version0.44.md:84]: Changing the title may be good enough.

Comment (pshipton) [reply docs/version0.44.md:84]: > Usually, the strategy that we use is to explain the problem first and then explain the solution or the change that was made to mitigate that problem.

Why do we do it this way? Perhaps it should be reversed. It seems more important to describe the change first with the explanation afterwards, since the explanation is optional.

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

File: docs/version0.44.md (hunk 2 above)
Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:86]: It is propounded in technical writing that users prefer to know what the problem was and then read about the solution for it. If we had a separate topic to give this explanation, then I could have given a generic explanation in this What's new topic. And like you had suggested, I have changed the title to be more specific about the change, so I guess that should address your concern? Thanks!

**Comment by Sreekala-Gopakumar** [comment]
@pshipton @keithc-ca - Any other changes required? Can this be merged? Thanks!

File: docs/version0.44.md [hunk 3]
```diff
@@ -76,6 +77,14 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
 
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Flag names that refer to the fields of `J9BuildFlags` changed in the DDR code
+
+The Direct Dump Reader (DDR) code enables the VM to read the system dump data by using the OpenJ9 Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool. The DDR code provides many built-in commands. You can extend the DDR code, adding your own commands, by writing plug-ins. These user plug-ins might contain code that has `J9BuildFlags` reference to read the build flags in the system dump data.
```
Comment (keithc-ca) [inline docs/version0.44.md:82]: This isn't really about "the VM"; suggest "enables reading system dump data" ...
reference -> references

File: docs/version0.44.md [hunk 4]
```diff
@@ -76,6 +77,14 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
...
+
+When the `J9BuildFlags` is generated for each build, the raw name of the build flags might be used in the plug-in code rather than the `cname` that appears in `j9cfg.h` (derived from `j9cfg.h.ft1`). Therefore, mapping of flag names in the plug-in code with the names that are specified in `j9cfg.h` was inconsistent. Introducing new flags meant either extra complexity or that some flag tests might not apply to all system dumps (because the flag name in `J9BuildFlags` was different at different times).
```
Comment (keithc-ca) [inline docs/version0.44.md:84]: Code that "might use" a raw name won't compile.
The root source file is `j9cfg.h.ftl` (the last letter is a lowercase `L`, not a `1`).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: docs/version0.44.md (hunk 4 above)
Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:84]: Regarding your first comment "Code that "might use" a raw name won't compile.", where exactly am I supposed to make the change? Is it the last sentence?

Also, is there any other term that we can use for "raw name"? What exactly is meant by "raw name of the build flags"?

Thanks!

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@keithc-ca - I have updated the content. Please check and let me know if any further changes are required. Thanks!

File: docs/version0.44.md [hunk 5]
```diff
@@ -31,6 +31,7 @@ The following new features and notable changes since version 0.43.0 are included
 - [XL C++ Runtime 16.1.0.7 or later required for AIX OpenJ9 builds on OpenJDK 8](#xl-c-runtime-16107-or-later-required-for-aix-openj9-builds-on-openjdk-8)
 - [New `-XX:[+|-]ShowUnmountedThreadStacks` option added](#new-xx-showunmountedthreadstacks-option-added)
 - [VMID query in the `jcmd` tool enhanced](#vmid-query-in-the-jcmd-tool-enhanced)
+- [ Flag names that refer to the fields of `J9BuildFlags` changed in the DDR code](#flag-names-that-refer-to-the-fields-of-j9buildflags-changed-in-the-ddr-code)
```
Comment (keithc-ca) [inline docs/version0.44.md:34]: No space after `[` for consistency with other entries.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
There are places where "the" appears that seem awkward or inappropriate, but overall I think  this communicates the necessary information.

File: docs/version0.44.md [hunk 6]
```diff
@@ -76,6 +77,16 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
 
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Flag names that refer to the fields of `J9BuildFlags` changed in the DDR code
+
+The Direct Dump Reader (DDR) code enables reading system dump data by using the OpenJ9 Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool. The DDR code provides many built-in commands and has flag names that refer to `J9BuildFlags` to read the build flags in the system dump data.
```
Comment (pshipton) [inline docs/version0.44.md:82]: These two things are not really related.

The DDR code provides many built-in commands and has flag names that refer to `J9BuildFlags` to read the build flags in the system dump data.

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@keithc-ca @pshipton  - I have updated the content. Please check and confirm. Thanks!

**Review by pshipton — APPROVED** [review]

File: docs/version0.44.md [hunk 7]
```diff
@@ -76,6 +77,16 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
...
+
+Earlier, some flag names in `J9BuildFlags` were mapped back to the flags in `j9.flags`. Such a mapping made providing support to all system dumps complex because some flag names in `J9BuildFlags` were different at different times. Now, when `J9BuildFlags` is generated for each build, the flag names are those names that are specified in `j9cfg.h` (derived from `j9cfg.h.ftl` or `j9cfg.h.in`) instead of the names that are defined in `j9.flags`. Code that has references to flag names that are defined in `j9.flags` results in compilation errors.
+
+Therefore, the flag names that are used in the DDR code to refer to the fields of `J9BuildFlags` are changed. The flag names are now those names that are in `j9cfg.h`. For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`.
+
+You can extend the DDR code, adding your own commands, by writing plug-ins. If the user plug-ins contain code that has `J9BuildFlags` references to read the build flags in the system dump data, you must use the names that are specified in `j9cfg.h` as flag names in the plug-in code also.
+
```
Comment (keithc-ca) [inline docs/version0.44.md:89]: Suggest rewriting this completely:
```
### Field names in `J9BuildFlags` have changed

The Direct Dump Reader (DDR) code enables reading system dump data by using the OpenJ9
Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool.
To facilitate access to build flags in the system dump data, DDR code uses fields of `J9BuildFlags`.
The names of those fields have varied over time making it challenging to support all system dumps.

Earlier, field names in `J9BuildFlags` were based on names defined in `j9.flags`.
Now, when the `J9BuildFlags` is generated for each build, the flag names are those
names that are specified in `j9cfg.h` (derived from `j9cfg.h.ftl` or `j9cfg.h.in`)
instead of the names that are defined in `j9.flags`.
For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`.

You can extend the DDR code, adding your own commands, by writing plug-ins.
If the user plug-in code contains references to fields of `J9BuildFlags` to read the build flags
in the system dump data, those references must change to use the names as specified in `j9cfg.h`.
```

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

File: docs/version0.44.md [hunk 8]
```diff
@@ -31,6 +31,7 @@ The following new features and notable changes since version 0.43.0 are included
 - [XL C++ Runtime 16.1.0.7 or later required for AIX OpenJ9 builds on OpenJDK 8](#xl-c-runtime-16107-or-later-required-for-aix-openj9-builds-on-openjdk-8)
 - [New `-XX:[+|-]ShowUnmountedThreadStacks` option added](#new-xx-showunmountedthreadstacks-option-added)
 - [VMID query in the `jcmd` tool enhanced](#vmid-query-in-the-jcmd-tool-enhanced)
+- [Field names in `J9BuildFlags` changed](#field-names-in-j9buildflags-changed)
```
Comment (pshipton) [inline docs/version0.44.md:34]: I think this should mention DDR so that someone with a DDR extension will know to read about the change.

File: docs/version0.44.md [hunk 9]
```diff
@@ -76,6 +77,17 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
...
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Field names in `J9BuildFlags` changed
+
+The Direct Dump Reader (DDR) code enables reading system dump data by using the OpenJ9
+Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool.
+DDR code uses fields of `J9BuildFlags` to access build flags in the system dump data.
+The names of `J9BuildFlags` fields changed over time because of which supporting different system dumps with different build flags became a challenge.
+
+Earlier, field names in `J9BuildFlags` were based on names defined in `j9.flags`. Now, when the `J9BuildFlags` is generated for each build, the flag names are those names that are specified in `j9cfgh` (derived from `j9cfg.h.ftl` or `j9cfg.h.in`) instead of the names that are defined in `j9.flags`. For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`.
```
Comment (pshipton) [inline docs/version0.44.md:87]: `j9cfgh` should be `j9cfg.h`

File: docs/version0.44.md [hunk 10]
```diff
@@ -76,6 +77,17 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
 
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Field names in `J9BuildFlags` changed
+
+The Direct Dump Reader (DDR) code enables reading system dump data by using the OpenJ9
+Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool.
+DDR code uses fields of `J9BuildFlags` to access build flags in the system dump data.
+The names of `J9BuildFlags` fields changed over time because of which supporting different system dumps with different build flags became a challenge.
```
Comment (keithc-ca) [inline docs/version0.44.md:85]: This sentence feels awkward; perhaps some punctuation will help and maybe remove the first occurrence of "different"?
What was your objection to my suggestion "making it challenging to support all system dumps"?

Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:85]: Even though I understand the sentence but the problem was with "... over time making it challenging...".  a noun followed by an "ing" as per Acrolinx. There is a possibility that the translation will find this confusing... is it "time making"? What is "it"? Also, leaving the sentence with "all system dumps". It was not clear what was challenging.

I will delete the first instance of "different". I hope that is ok.

Comment (keithc-ca) [reply docs/version0.44.md:85]: I expect punctuation (a comma) would have been helpful in my suggestion
> fields changed over time, making it challenging ...

I think it would be a similar improvement to your proposal
> fields changed over time, because of which ...

In both cases it is the "change" that makes it challenging, or it is because of the "change". I don't how we've avoided the problem you describe.

Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:85]: How about "The names of `J9BuildFlags` fields changed over time and therefore, supporting system dumps with different build flags became a challenge."?

Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:85]: Also, because of Peter's suggestion about mentioning DDR code in the heading, I am also adding a sentence at the end of the second para.

> Earlier, field names in `J9BuildFlags` were based on names defined in `j9.flags`. Now, when the `J9BuildFlags` is generated for each build, the flag names are those names that are specified in `j9cfg.h` (derived from `j9cfg.h.ftl` or `j9cfg.h.in`) instead of the names that are defined in `j9.flags`. For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`. Because of this change, reference to the field names in `J9BuildFlags` is changed in the DDR code as well.

Is that ok?

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Comment by Sreekala-Gopakumar** [comment]
@keithc-ca - Please check. Thanks!

File: docs/version0.44.md [hunk 11]
```diff
@@ -31,6 +31,7 @@ The following new features and notable changes since version 0.43.0 are included
 - [XL C++ Runtime 16.1.0.7 or later required for AIX OpenJ9 builds on OpenJDK 8](#xl-c-runtime-16107-or-later-required-for-aix-openj9-builds-on-openjdk-8)
 - [New `-XX:[+|-]ShowUnmountedThreadStacks` option added](#new-xx-showunmountedthreadstacks-option-added)
 - [VMID query in the `jcmd` tool enhanced](#vmid-query-in-the-jcmd-tool-enhanced)
+- [Reference to the field names in `J9BuildFlags` changed in the DDR code](#reference-to-the-field-names-in-j9buildflags-changed-in-the-ddr-code)
```
Comment (keithc-ca) [inline docs/version0.44.md:34]: Besides the grammar problem (it would say "references", not "reference"), this isn't a fair summary; suggest: "DDR field names in J9BuildFlags changed".

File: docs/version0.44.md [hunk 12]
```diff
@@ -76,6 +77,17 @@ The `jcmd` tool also now supports specifying `0` as a VMID to target all VMs.
...
 For more information, see [Java diagnostic command (`jcmd`) tool](tool_jcmd.md).
 
+### Reference to the field names in `J9BuildFlags` changed in the DDR code
+
+The Direct Dump Reader (DDR) code enables reading system dump data by using the OpenJ9
+Diagnostic Tool Framework for Java (DTFJ) API or the [`jdmpview`](tool_jdmpview.md) tool.
+DDR code uses fields of `J9BuildFlags` to access build flags in the system dump data.
+The names of `J9BuildFlags` fields changed over time and therefore, supporting system dumps with different build flags became a challenge.
+
+Earlier, field names in `J9BuildFlags` were based on names defined in `j9.flags`. Now, when the `J9BuildFlags` is generated for each build, the flag names are those names that are specified in `j9cfg.h` (derived from `j9cfg.h.ftl` or `j9cfg.h.in`) instead of the names that are defined in `j9.flags`. For example, `env_data64` is now referred to as `J9VM_ENV_DATA64`. Because of this change, reference to the field names in `J9BuildFlags` is changed in the DDR code as well.
```
Comment (keithc-ca) [inline docs/version0.44.md:87]: The final sentence here is not helpful (and is also grammatically incorrect).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: docs/version0.44.md (hunk 12 above)
Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:87]: I need a sentence to link the change in `J9BuildFlags` to the corresponding change in the DDR code. Aren't `J9BuildFlags` and DDR code two different things? Trying to get the concept clear here.

Comment (Sreekala-Gopakumar) [reply docs/version0.44.md:87]: Yes, I understood that from your detailed explanation in one of your earlier comments. What I am trying to do here is create a correct flow between sentences. So, in my understanding, there are 3 steps - the names were first changed in `J9BuildFlags` (which is different from DDR code), then because of this change, the DDR code that has references to `J9BuildFlags` was changed and ultimately the expectation is that the plug-in code should also change. Is that correct?

My confusion was because of the following line:
> DDR code uses fields of `J9BuildFlags` to access build flags in the system dump data

whether when you say "...uses fields of `J9BuildFlags`" then do you mean that 

- "fields of `J9BuildFlags` is separate from DDR code and the DDR code has  flag names that are used to access those fields, or 
- that actually `J9BuildFlags` is not separate and is a part of DDR code and there are only 2 steps - one,  the DDR code that includes `J9BuildFlags` was changed and therefore the plug-in code should also change.

Comment (keithc-ca) [reply docs/version0.44.md:87]: From that perspective, only one thing has changed: the API that DDR plug-in code might use. The rest of the text explains why the change was made.

**Comment by Sreekala-Gopakumar** [comment]
@keithc-ca - I have deleted the line "Because of this change, reference to the field names in `J9BuildFlags` is changed in the DDR code as well." I hope this is ok. Can this be finalised now? Thanks!

**Comment by Sreekala-Gopakumar** [comment]
Jenkins doc stage

**Review by pshipton — APPROVED** [review]
